systems/controllers/osc/operational_space_control.cc

```diff
@@ -396,33 +418,42 @@ VectorXd OperationalSpaceControl::SolveQp(
   bias = bias - grav;
 
   // Get J and JdotV for holonomic constraint
-  MatrixXd J_h(distance_constraints_.size(), n_v_);
-  VectorXd JdotV_h(distance_constraints_.size());
-  for (unsigned int i = 0; i < distance_constraints_.size(); ++i) {
-    distance_constraints_[i]->updateConstraint(*context_wo_spr_);
-    J_h.row(i) = distance_constraints_[i]->getJ();
-    JdotV_h.segment(i, 1) = distance_constraints_[i]->getJdotv();
+  MatrixXd J_h(n_h_, n_v_);
+  VectorXd JdotV_h(n_h_);
+  if (kinematic_evaluators_ != nullptr) {
+    J_h = kinematic_evaluators_->EvalFullJacobian(*context_wo_spr_);
+    JdotV_h =
+        kinematic_evaluators_->EvalFullJacobianDotTimesV(*context_wo_spr_);
   }
 
-  // Get J and JdotV for contact constraint
+  // Get J for external forces in equations of motion
   MatrixXd J_c = MatrixXd::Zero(n_c_, n_v_);
-  VectorXd JdotV_c = VectorXd::Zero(n_c_);
-  for (unsigned int i = 0; i < active_contact_flags.size(); i++) {
-    if (active_contact_flags[i]) {
-      MatrixXd J = MatrixXd::Zero(3, n_v_);
-      plant_wo_spr_.CalcJacobianTranslationalVelocity(
-          *context_wo_spr_, JacobianWrtVariable::kV,
-          plant_wo_spr_.get_body(body_indices_[i]).body_frame(),
-          pts_on_body_[i], world_wo_spr_, world_wo_spr_, &J);
-      J_c.block(3 * i, 0, 3, n_v_) = J;
-      JdotV_c.segment(3 * i, 3) =
-          plant_wo_spr_
-              .CalcBiasForJacobianSpatialVelocity(
-                  *context_wo_spr_, JacobianWrtVariable::kV,
-                  plant_wo_spr_.get_body(body_indices_[i]).body_frame(),
-                  pts_on_body_[i], world_wo_spr_, world_wo_spr_)
-              .tail(3);
+  for (unsigned int i = 0; i < all_contacts_.size(); i++) {
+    if (std::find(active_contact_list.begin(), active_contact_list.end(), i) !=
+        active_contact_list.end()) {
+      J_c.block(3 * i, 0, 3, n_v_) =
```

Comment:
although it may be obvious, we should use a constant instead of the number 3. Maybe name the constant `static const int SPACE_DIM` or something similar

Reply:
Done